File: backend/src/main/java/com/eai/application/tenant/CompanyService.java
package com.eai.application.tenant;

import com.eai.application.common.NotFoundException;
import com.eai.application.user.UserRepository;
import com.eai.domain.tenant.Company;
import com.eai.domain.tenant.TenantStatus;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class CompanyService {

    private final CompanyRepository companyRepository;
    private final UserRepository userRepository;

    @Transactional(readOnly = true)
    public List<Company> listCompanies() {
        return companyRepository.findAll();
    }

    @Transactional(readOnly = true)
    public Company getCompany(UUID id) {
        return findRequired(id);
    }

    @Transactional
    public Company createCompany(CreateCompanyCommand command) {
        return companyRepository.save(Company.create(command.name()));
    }

    @Transactional
    public Company updateCompany(UUID id, UpdateCompanyCommand command) {
        Company company = findRequired(id);
        if (company.getStatus() != TenantStatus.INACTIVE
                && command.status() == TenantStatus.INACTIVE
                && userRepository.existsActiveByCompanyId(id)) {
            throw new IllegalArgumentException("Cannot deactivate company with active users");
        }
        company.update(command.name(), command.status());
        return companyRepository.save(company);
    }

    public Company findRequired(UUID id) {
        return companyRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Company not found"));
    }

}

File: frontend/src/services/companyService.ts
import { api } from './api';
import type { Company, TenantStatus } from '../types/tenant';

export interface CompanyPayload {
  name: string;
  status?: TenantStatus;
}

export interface UpdateCompanyPayload extends CompanyPayload {
  status: TenantStatus;
}

export async function listCompanies() {
  const response = await api.get<Company[]>('/companies');
  return response.data;
}

export async function createCompany(payload: CompanyPayload) {
  const response = await api.post<Company>('/companies', payload);
  return response.data;
}

export async function updateCompany(id: string, payload: UpdateCompanyPayload) {
  const response = await api.put<Company>(`/companies/${id}`, payload);
  return response.data;
}
